GVFS/GVFS.UnitTests/Maintenance/GitMaintenanceQueueTests.cs

```diff
@@ -0,0 +1,200 @@
+            {
+                enlistment.EnlistmentRoot,
+                enlistment.GitObjectsRoot
+            });
+
+            GVFSContext context = new GVFSContext(tracer, fileSystem, null, enlistment);
+            GitObjects gitObjects = new MockPhysicalGitObjects(tracer, null, null, null);
+
+            TestGitMaintenanceStep step1 = new TestGitMaintenanceStep(context, gitObjects);
+            TestGitMaintenanceStep step2 = new TestGitMaintenanceStep(context, gitObjects);
+
+            GitMaintenanceQueue queue = new GitMaintenanceQueue(context);
+
+            queue.Enqueue(step1);
+            queue.Enqueue(step2);
+
+            Assert.IsTrue(step1.EventTriggered.WaitOne(this.maxWaitTime)
+                && step2.EventTriggered.WaitOne(this.maxWaitTime));
+
+            queue.Stop();
+        }
+
+        [TestCase]
+        public void GitMaintenanceQueueStopSuceedsWhenQueueIsEmpty()
+        {
+            ITracer tracer = new MockTracer();
+            GVFSEnlistment enlistment = new MockGVFSEnlistment();
+
+            // We need to have the EnlistmentRoot and GitObjectsRoot available for jobs to run
+            ReadyFileSystem fileSystem = new ReadyFileSystem(new string[]
+            {
+                enlistment.EnlistmentRoot,
+                enlistment.GitObjectsRoot
+            });
+
+            GVFSContext context = new GVFSContext(tracer, fileSystem, null, enlistment);
+            GitObjects gitObjects = new MockPhysicalGitObjects(tracer, null, null, null);
+
+            GitMaintenanceQueue queue = new GitMaintenanceQueue(context);
+
+            queue.Stop();
+        }
+
+        [TestCase]
+        public void GitMaintenanceQueueStopsJob()
+        {
+            ITracer tracer = new MockTracer();
+            GVFSEnlistment enlistment = new MockGVFSEnlistment();
+
+            // We need to have the EnlistmentRoot and GitObjectsRoot available for jobs to run
+            ReadyFileSystem fileSystem = new ReadyFileSystem(new string[]
+            {
+                enlistment.EnlistmentRoot,
+                enlistment.GitObjectsRoot
+            });
+
+            GVFSContext context = new GVFSContext(tracer, fileSystem, null, enlistment);
+            GitObjects gitObjects = new MockPhysicalGitObjects(tracer, null, null, null);
+
+            GitMaintenanceQueue queue = new GitMaintenanceQueue(context);
+
+            // This step stops the queue after the step is started,
+            // then checks if Stop() was called.
+            WatchForStopStep watchForStop = new WatchForStopStep(queue, context, gitObjects);
+
+            queue.Enqueue(watchForStop);
+            Assert.IsTrue(watchForStop.EventTriggered.WaitOne(this.maxWaitTime));
+            watchForStop.SawStopping.ShouldBeTrue();
+
+            // Ensure we don't start a job after the Stop() call
+            TestGitMaintenanceStep watchForStart = new TestGitMaintenanceStep(context, gitObjects);
+            queue.Enqueue(watchForStart);

```

Comment:
Should this throw an exception or have a return value that we can check if it is something that isn't allowed?

Reply:
As long as we keep in mind that product code doesn't need to check for the return value. As far as the process is concerned, this is a noisy channel and steps can be dropped with no warning (because the lock is taken, because the drive is unavailable, because someone already called `Stop()` on the queue, etc.). Since we've encapsulated the queue in the scheduler, this surface is very small.